### PR TITLE
Remove iteration counter and norms that are already covered by SolverStatistics

### PR DESCRIPTION
**Pull request by pschultzendorff**
Changes proposed by @IngridKJ and I. 

## Proposed changes

Adresses issues #1219 and #1200.

``model.nonlinar_solver_statistics`` already keeps track of the number of nonlinear iterations, residual norms, and nonlinear increment norms. Thus, any explicit passes of these values between ``SolutionStrategy.after_nonlinear_convergence``, ``SolutionStrategy.check_convergence``, and ``NewtonSolver.solve`` are removed. Conveniently, this fixes ``iteration_counter`` lagging behind by one step. ``NewtonSolver.solve`` only returns one boolean value ``converged`` now.

On an additional note, we noticed that the test suite does not cover the case of divergence due to nan values in ``SolutionStrategy.check_convergence``, i.e., lines 547-551 in ``solution_strategy.py``. We only noticed a mistake in our changes due to flake complaining, not the tests. Any opinions on that would be welcome :)

We added tests for ``SolutionStrategy.check_convergence`` and ``NewtonSolver.solve``.

## Types of changes

What types of changes does this PR introduce to PorePy?

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

Not quite sure how to categorize this, as it is only code cleanup, but any custom scripts depending on ``iteration_counter`` and ``check_convergence`` will break.

## Checklist

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/numerics/nonlinear/nonlinear_solvers.py [hunk 1]
```diff
@@ -145,19 +134,23 @@ def newton_step() -> None:
...
+                        + f"{model.nonlinear_solver_statistics.num_iteration + 1} of \t"
+                        + f"{self.params['max_iterations']}"
+                        ""
                     )
                     newton_step()
-                    iteration_counter += 1
 
                     solver_progressbar.update(n=1)
+                    # Ignore line being too long, because we would need an additional
+                    # variable to fix this.
```
Comment (IvarStefansson) [inline src/porepy/numerics/nonlinear/nonlinear_solvers.py:151]: Do you agree with this, @keileg?

File: src/porepy/numerics/nonlinear/nonlinear_solvers.py [hunk 2]
```diff
@@ -168,15 +161,13 @@ def newton_step() -> None:
...
-                            iteration_counter=iteration_counter
-                        )
+                        model.after_nonlinear_convergence()
                         break
 
         if not is_converged:
             model.after_nonlinear_failure()
 
-        return is_converged, iteration_counter
+        return is_converged, model.nonlinear_solver_statistics.num_iteration
```
Comment (IvarStefansson) [inline src/porepy/numerics/nonlinear/nonlinear_solvers.py:170]: Is it important to return the counter?

**Review by IvarStefansson — COMMENTED** [review]
Thanks!
Would it make sense to add a test that runs for a fixed number of iterations and than e.g. checks that at the end the iteration count is as expected?

File: src/porepy/numerics/nonlinear/nonlinear_solvers.py (hunk 2 above)
Comment (pschultzendorff) [reply src/porepy/numerics/nonlinear/nonlinear_solvers.py:170]: The only uses of ``NewtonSolver.solve`` in PorePy are in ``run_models.py``, which
ignores both return values. We can freely choose what to return. Although, for a
function that is called solve, having some return value other than ``None`` is
probably good practice.

**Comment by pschultzendorff** [comment]
> Would it make sense to add a test that runs for a fixed number of iterations and than e.g. checks that at the end the iteration count is as expected?

Since the issues were about the iteration count being off by one, I like the idea! Will add a test for this.

File: src/porepy/numerics/nonlinear/nonlinear_solvers.py (hunk 2 above)
Comment (IvarStefansson) [reply src/porepy/numerics/nonlinear/nonlinear_solvers.py:170]: Yes, I'd keep is_converged. But I think we should remove the counter, which can be retreived from the model's solution statistic object.

File: src/porepy/numerics/nonlinear/nonlinear_solvers.py (hunk 1 above)
Comment (keileg) [reply src/porepy/numerics/nonlinear/nonlinear_solvers.py:151]: Yes. Sorry for the late response.

**Comment by pschultzendorff** [comment]
We addressed the comments and added tests.

File: tests/numerics/nonlinear/test_nonlinear_solvers.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            < self.expected_number_of_iterations
+        ):
+            converged = False
+            return converged, diverged
+        converged = True
+        return converged, diverged
+
+    def after_nonlinear_failure(self) -> None:
+        """Method which is called if the non-linear solver fails to converge."""
+        self.save_data_time_step()
```
Comment (IvarStefansson) [inline tests/numerics/nonlinear/test_nonlinear_solvers.py:41]: Why is this needed?

File: tests/numerics/nonlinear/test_nonlinear_solvers.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+from porepy.models.fluid_mass_balance import SinglePhaseFlow
+
+
+class NonlinearSinglePhaseFlow(SinglePhaseFlow):
+    """Class setup which forces a set number of nonlinear iterations to be run for
+    testing."""
+
+    def _is_nonlinear_problem(self):
+        """Method for stating that the problem is nonlinear."""
+        return True
```
Comment (IvarStefansson) [inline tests/numerics/nonlinear/test_nonlinear_solvers.py:14]: Is this not true by default?

File: tests/numerics/nonlinear/test_nonlinear_solvers.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+def test_nonlinear_iteration_count():
+    """Test for checking if the nonlinear iterations are counted as expected.
+
+    A pre set value of expected iterations is set, and the test checks that the
+    iteration count matches the pre set value after convergence is obtained.
+
+    """
+    params = {"max_iterations": 96}
+    model = NonlinearSinglePhaseFlow()
+    model.expected_number_of_iterations = 13
```
Comment (IvarStefansson) [inline tests/numerics/nonlinear/test_nonlinear_solvers.py:55]: Can't we make this number smaller to avoid unnecessary test time?

**Review by IvarStefansson — COMMENTED** [review]
Couple of small questions to the new test.

File: tests/numerics/nonlinear/test_nonlinear_solvers.py (hunk 3 above)
Comment (IngridKJ) [reply tests/numerics/nonlinear/test_nonlinear_solvers.py:41]: Hm, apparently it is not needed. I had some error messages when I made test, and including this method (seemingly) removed the problem. I see now that it is not needed, so I will delete it.

**Review by jwboth — APPROVED** [review]
Thanks both of you for removing this redundancy and adding a test!
